### ISSUE TITLE
Support for drums/percussion?

### ISSUE THREAD

**Issue by tripathiarpan20** [spotify/basic-pitch#30]
Hi!
Thanks for this amazing open-source work, I'm really enjoying using it. :)

I noticed that Basic Pitch works great for tracks with single mono/polyphonic instrument for most instruments, however it is unable to encode drums at all. 

I understand that MIDI encoding for drums/percussion instruments is somewhat different compared to the rest of the instruments, but are there any future plans to add support for percussion instruments?

**Comment by jugoodma** [comment]
@tripathiarpan20 -- I found your comment interesting, so I took a short dive into the literature.

There's a niche, and interesting, sub-sub-field of Music Information Retrieval (MIR) called Automatic Drum Transcription (ADT). [Here's a literature review](http://www.open-access.bcu.ac.uk/6180/1/Wu-et-al.-2018-A-review-of-automatic-drum-transcription.pdf) of ADT. The authors of that review describe different "drum transcription tasks" -- drum-only transcription (DTD) and drum-plus-accompaniment transcription (DTM) seem particularly relevant.

If you want to "solve" drum encoding, you could look at some of the methods in the more recently referenced papers in the mentioned literature review and give them a try! [Ref 80](http://www.open-access.bcu.ac.uk/4747/1/Automatic%20drum%20transcription%20for%20polyphonic.pdf) appeared to have high scoring metrics, but might not work for drum kits with more than a kick, snare, and hi-hat. The authors (of ref 80) also have a [GitHub](https://github.com/CarlSouthall/ADTLib) repo, and a demo site linked!

For another approach, you might find <https://github.com/magenta/mt3> interesting/useful. Unfortunately, the related paper doesn't focus too heavily on drums, so you might find the `mt3` model doesn't work _that well_ for drum transcription.

Finally, perhaps we could make use of Facebook's [`demucs`](https://github.com/facebookresearch/demucs). This model is seemingly SOTA for demixing audio tracks, so we can use it to separate out the drums stem of a track. This turns a DTM task into a DTD task quite effectively (and thus, in my opinion, makes solving ADT _easier_). Unfortunately, this somewhat disregards the call-to-action in the `NMP`/`basic-pitch` paper -- to encourage low-resource models in future research. Maybe we can trim down the `demucs` model? Regardless, perhaps we could then train the `NMP` model on a drum-specific dataset, like [E-GMD](https://magenta.tensorflow.org/datasets/e-gmd). We could then compose the architectures like so:

```txt
                demucs                   NMP(E-GMD)
original track -------> drum-only track -----------> drum-only MIDI
```

I'll give this a try, and post on the results. Luckily, since NMP is so light it _probably_ trains much faster than huge models, And who knows, maybe `demucs` isn't even needed. Or, maybe this entire approach won't work! It's all part of the scientific method 😄

**Comment by rabitt** [comment]
> are there any future plans to add support for percussion instruments?

@tripathiarpan20 no plans at the moment, but will let you know if that changes. @jugoodma 's comment is great, and points to some open source drum transcription options. Here are two more open source systems I'm aware of: 
(1) [paper](https://markcartwright.com/files/cartwright2018increasing.pdf) [code](https://github.com/mcartwright/dafx2018_adt)
(2) [paper](https://arxiv.org/pdf/1806.06676.pdf) [code](http://ifs.tuwien.ac.at/~vogl/dafx2018/)

**Comment by tripathiarpan20** [comment]
Hi @jugoodma and @rabitt ,
Thank you for the amazing feedbacks!

To be frank I am not familiar with how the instrument class is predicted in the NMP pipeline, but if retraining the Basic Pitch's architecture on Drum dataset for DTD along with devising the suitable posteriorgram post-processing works, I believe that it would make the domain of instruments in this project truly whole (afaik).  

Good luck on the process and keep us updated :D. 
The DTD task seems to be the relevant one in the context of Basic Pitch (which deals with polyphonic recordings of a single instrument class), `demucs` shouldn't be required given its high inference time and the availability of the E-GMD dataset & conversion to drum audio tracks with suitable soundfonts and label preserving data augmentation. 

Elsewhere, I also tried `demucs` on Psychosocial(Slipknot) & tried to use basic-pitch on the demixed drum track, and that's how I eventually raised the issue/question. Although `demucs` has amazing performance, the inference times are relatively higher (typically takes minutes). 

Meanwhile, perhaps Spotify could develop a lightweight demixing model which might benefit from end-to-end deep learning that uses CQT for preprocessing (rather than Mel spectrograms as in past demixing methods) in the future? 
It might be bit of a stretch as my understanding of the working of spectrograms, past Demixing models & NMP has missing pieces. 
I would especially like to hear @rabitt 's thoughts on the feasibility of such a lightweight demixing model and whether there would be any benefits if it is formulated as an end-to-end (demixing + transcription) task. 

Any feedback from anyone else is welcome too!

**Comment by sevagh** [comment]
Is the original problem that this model cannot transcribe drums _within a fully mixed song_ with other instruments present? What happens if you feed it an isolated drum stem?

**Comment by jugoodma** [comment]
@sslupsky I tried it out when I posted my original comment, but I couldn't get the training to work locally (to be fair, I didn't try that hard) (or maybe I did get it working, but the training was taking a while so I thought I had the setup wrong?). At that time, the full training pipeline hadn't been committed to this repo -- from what I remember, though, the training pipeline is published now? If I ever find myself with some time, I'll give the training another go :)

@sevagh from what I remember, the original proposal was to train NMP with (isolated drum samples, corresponding MIDI). Assuming this works, then it seems reasonable to use @sslupsky's pipeline proposal if you wanted exclusively a drum MIDI from a full song (sample -> demucs(drum-stem) -> NMP(E-GMD))

**Comment by sslupsky** [comment]
@jugoodma Have you seen RipX?  I stumbled across this software a while ago and had a look.  It appears to do something like this for all the STEMS but theirs is commercial software.

**Comment by sevagh** [comment]
I think two issues are being conflated here. Demucs is of course a great model for source separation, but the paper for Basic Pitch barely mentions percussion. Everything in the paper and model points to being designed for pitched/harmonic instruments only.

So, Demucs might solve the problem of "how to get drum tracks" but it's putting the cart before the horse. I think you'll have trouble coercing a model that hasn't been designed for drum/percussion transcription to doing that task.

The paper talks about F0 for example as their main accuracy measure, and perucssion won't have a fundamental frequency the same way.

**Comment by bgenchel** [comment]
@sevagh is correct, the way this model is designed and trained is geared towards pitch / harmonic detection and classification and so training it on percussion only will likely not yield the desired effect.
